pkgs/applications/editors/emacs/build-support/melpa.nix

```diff
@@ -62,17 +64,17 @@ in
     recipe: Optional MELPA recipe.
     Default: a minimally functional recipe
   */
-, recipe ? (writeText "${pname}-recipe" ''
-    (${ename} :fetcher git :url ""
-              ${lib.optionalString (files != null) ":files ${files}"})
+, recipe ? (writeText "${finalAttrs.pname}-recipe" ''
+    (${finalAttrs.ename} :fetcher git :url ""
+              ${lib.optionalString (finalAttrs.files or null != null) ":files ${finalAttrs.files}"})
```

Comment:
One must not attempt to determine the existence of an attribute in `finalAttrs`, as this will lead to infinite recursion when evaluated.

The CI is right about the arguments `files` being unused. We need to pass `file` down the base build helper to ensure its existence, and then we can access `finalAttrs`.

Reply:
> One must not attempt to determine the existence of an attribute in `finalAttrs`, as this will lead to infinite recursion when evaluated.

Both packages with null `files` (`emacs.pkgs.git-undo`) and non-null `files` (`emacs.pkgs.urweb-mode`) build.  Also, ofborg evaluates without error.  So presumably there is no infinite recursion here.

Reply:
> Both packages with null `files` (`emacs.pkgs.git-undo`) and non-null `files` (`emacs.pkgs.urweb-mode`) build.

The error is not shown simply because the default value of `recipe` is not evaluated.

Try to pass the default value (e.g., under `passthru`) and access it, and you'll see the `infinite recursion` error.

Reply:
> The error is not shown simply because the default value of `recipe` is not evaluated.

I beg to differ.

The build process actually [uses][1] `recipe`.

Building a recipe also works.

```console
> cat (nix build --include nixpkgs=$PWD --file . emacs.pkgs.git-undo.recipe --print-out-paths)
(git-undo :fetcher git :url ""
          )
```

Here is a minimal working example to show this pattern works:

```console
nix-repl> f = self: {x = self.y or 666; }                                         

nix-repl> (lib.fix f).x
666

nix-repl> 
```

[1]: https://github.com/NixOS/nixpkgs/blob/0f44621c076ad3dc27225de13894e5c36e8989a7/pkgs/applications/editors/emacs/build-support/melpa.nix#L95-L99

Reply:
That's strange. I'll go ahead and check it out.

Reply:
You're right. The pattern that would cause infinite recursion is "specifying the value of an attribute based on its existence in `finalAttrs`."